cognite/powerops/preprocessor/cogreader.py

```diff
@@ -2,22 +2,77 @@
 
 import tempfile
 from functools import cached_property
-from typing import List, Optional, Union
+from typing import List, Literal, Optional, TypedDict, Union
 
 from cognite.client import CogniteClient
+from pydantic import BaseModel, root_validator
 
 from cognite.powerops.preprocessor import knockoff_logging as logging
 from cognite.powerops.preprocessor.data_classes import CogShopCase, CogShopConfig
 from cognite.powerops.preprocessor.data_classes.time_series_mapping import TimeSeriesMapping
 from cognite.powerops.preprocessor.exceptions import CogReaderError
-from cognite.powerops.preprocessor.utils import ShopMetadata, download_file, log_and_reraise, retrieve_yaml_file
+from cognite.powerops.preprocessor.utils import ShopMetadata, download_file, log_and_reraise, now, retrieve_yaml_file
 
 from .get_fdm_data import Case, get_case
-from .utils import find_closest_file, group_files_by_metadata
+from .utils import find_closest_file
 
 logger = logging.getLogger(__name__)
 
 
+class CogShopFileDict(TypedDict):
+    external_id: str
+    file_type: str
+
+
+class SortBy(TypedDict):
+    metadata_key: Optional[str]
+    file_attribute: Optional[Literal["last_updated_time", "created_time", "uploaded_time"]]
+
+
+class CogShopFile(BaseModel):
+    label: Optional[str]
+    pick: Optional[Literal["closest", "exact", "latest"]]
+    sort_by: Optional[SortBy]
+    external_id_prefix: Optional[str]
+    external_id: Optional[str]
+    file_type: Literal["ascii", "yaml"]
+
+    class Config:
+        validate_all = True
+
+    @root_validator
+    def external_id_or_prefix_set(cls, values):
+        if not values.get("external_id") and not values.get("external_id_prefix"):
+            raise ValueError("Either external_id or external_id_prefix must be set")
+        return values
+
+    def get_file_dict(self, client: CogniteClient, starttime_ms: float) -> CogShopFileDict:
+        if self.external_id:
+            return {"external_id": self.external_id, "file_type": self.file_type}
+        elif self.external_id_prefix and self.pick == "closest":
+            files = client.files.list(external_id_prefix=self.external_id_prefix, limit=None)
+            if closest_file := find_closest_file(files, starttime_ms, self.sort_by):
```

Comment:
mypy gives a type error here. But importing SortBy to utils causes a circular dependency. Any tips here? maybe I'm overcomplicating stuff, but would be good to have to Literal[] type check for the file_attribute. Or I could make it to a pydantic class and create a validator -wdyt?